{% import 'utils.html' as utils %}
{% extends 'base.html' %}

{% block header %}
<style media="screen">
@media (max-width: 575.98px) {
  #legend {
    width: 100%;
  }
}

</style>

{% endblock %}

{% block content %}
	<div class="row text-center mx-auto">
		<div class="col-md-1">

		</div>
		<div class="col-md-10 pt-4">

			<img src="{{ url_for('static', filename='covid_profiler_logo.svg') }}" alt="" width="75%">
			<div class="pb-4 text-justify">
				Welcome to the COVID-Profiler - a pipeline which allows users
				to analyse <i>Sars-Cov-2</i> sequencing data to place isolates within
                a real-time tree.
			</div>



			<h5 class="">How does it work?</h5>
			<div class="pb-4 text-justify">
                This tree has been created using sequences from the
                <a href="https://www.gisaid.org/">GISAID database</a>. Users can
                upload their own sequences to look where they are positioned on the tree.
				The pipeline searches for SNPs associated with branches on the tree.
                It then uses the variants to find the closest neighbours on the tree.
                In parallel it classifies the strains into S/L types as defined by
                <a href="https://academic.oup.com/nsr/advance-article/doi/10.1093/nsr/nwaa036/5775463">Tang et al</a>.
			</div>



            <div class="row justify-content-md-center">
                <div class="col-md-10 pb-4">
                    <div class="card border-dark text-center">
                        <div class="card-header">
                            Phylogenetic tree
                        </div>
                            <div class="card-body" style="overflow: hidden;">
                                <div class="">
                                    <img src="{{ url_for('static', filename='tree_legend.svg') }}" id="legend" alt="" width="50%">
                                </div>
                                <div class="" style="">
                                    <svg id="tree_display"></svg>
                                </div>
                            </div>
                            <div class="card-footer text-muted">
                              Last updated: {{ tree["created"] }}
                            </div>
                    </div>
                </div>
            </div>



			<div class="row justify-content-md-center">
				<div class="col-md-6 pb-4">
					<div class="card border-dark text-center">
					  <div class="card-header">
					    Step 1
					  </div>
					  <div class="card-body">
					    <h5 class="card-title">Profile your sample</h5>

									<p class="card-text">
										Upload your sequencing data in <b>fasta</b> format.
									</p>


					    <a class="btn btn-primary" href="{{ url_for('upload.upload') }}">Upload</a>
					  </div>
					</div>
				</div>
			</div>


			<div class="row justify-content-md-center">
				<div class="col-md-6 pb-4">
					<div class="card border-dark text-center">
					  <div class="card-header">
					    Step 2
					  </div>
					  <div class="card-body">
					    <h5 class="card-title">View the results</h5>

									<p class="card-text">
										Find your results by entering your unique run ID directly into the search box below.
									</p>

							<form method="post" enctype=multipart/form-data>
								<div class="row justify-content-md-center">
									<div class="col-md-6">
										<input type="text" name="sample_id" class="form-control" placeholder="Sample ID">
									</div>
									<button type="submit" class="btn btn-primary">Submit</button>
								</div>
							</form>
					  </div>
					</div>
				</div>
			</div>

            <h5 class="">Get in touch</h5>
			<div class="text-left text-justify">
                This work was made possible through a great collaborative effort.
                We welcome any suggestions or contributions, so do not hesitate to
                get in touch with <a href="mailto:&#106;&#111;&#100;&#121;&#046;&#112;&#104;&#101;&#108;&#097;&#110;&#064;&#108;&#115;&#104;&#116;&#109;&#046;&#097;&#099;&#046;&#117;&#107;">Jody Phelan</a> or <a href="mailto:&#087;&#111;&#117;&#116;&#101;&#114;&#046;&#068;&#101;&#101;&#108;&#100;&#101;&#114;&#064;&#108;&#115;&#104;&#116;&#109;&#046;&#097;&#099;&#046;&#117;&#107;">Wouter Deelder</a>. The project is completely open source and the links to
                the source code is available below.

			</div>


            <hr>
			<h5 class="">Pipeline</h5>
			<div class="text-left text-justify">
                Sequences are aligned to the reference genome (NC_045512.2) using minimap2 [1].
                Variants are called using paftools.js from the minimap2 package and bcftools [2]
                and transformed into a fasta alignment. IQ-TREE [3] is used to reconstruct the
                phylogeny. Custom scripts utilising ete3 [4] are used to position new isolates
                on the tree which are visualised using phylotree.js [5] in the web browser.
                Sample data is downloaded from NCBI nucleotide database using <a href="https://www.ncbi.nlm.nih.gov/books/NBK179288/">Entrez Direct</a>.
                The website is built using <a href="https://github.com/pallets/flask">flask</a>.
                All the code is available from <a href="https://github.com/jodyphelan/covid-profiler">https://github.com/jodyphelan/covid-profiler</a>.
                <div class="pt-4 text-justify">
                    <ol>
                        <li>Li, H. Minimap2: Pairwise alignment for nucleotide sequences. <i>Bioinformatics</i> <b>34</b>, 3094–3100 (2018).</li>
                        <li>Li, H. A statistical framework for SNP calling, mutation discovery, association mapping and population genetical parameter estimation from sequencing data. <i>Bioinformatics</i> <b>27</b>, 2987–2993 (2011).</li>
                        <li>Minh, B. Q. et al. IQ-TREE 2: New models and efficient methods for phylogenetic inference in the genomic era. <i>Mol. Biol. Evol.</i> (2020). doi:10.1093/molbev/msaa015</li>
                        <li>Huerta-Cepas, J., Serra, F. & Bork, P. ETE 3: Reconstruction, Analysis, and Visualization of Phylogenomic Data. <i>Mol. Biol. Evol.</i> <b>33</b>, 1635–8 (2016).</li>
                        <li>Shank, S. D., Weaver, S. & Kosakovsky Pond, S. L. phylotree.js - a JavaScript library for application development and interactive data visualization in phylogenetics. <i>BMC Bioinformatics</i> <b>19</b>, 276 (2018).</li>
                    </ol>
                </div>
			</div>


			<hr>
			<h6 class="">Disclaimer</h6>
			<div class="text-justify">
				This tool is for Research Use Only. It has not been approved, cleared,
				or licensed by any regulatory authority. By submitting sequence data the
				 user acknowledges no intended medical purpose or objective such as
				 clinical diagnosis, patient management, or human clinical trials.
			</div>
		</div>
	</div>

    <script type="text/javascript">
    $(document).ready( function () {
        var data = '{{ tree["meta"] | safe }}'
        var tree_meta = JSON.parse(data)
        console.log(tree_meta)
        var newick_string = '{{ tree["newick"] }}'
        if (window.innerWidth>500){
            var height = window.innerWidth/4
        } else {
            var height = window.innerWidth/3
        }
        var width = $( "#tree_display" ).parent().width()-40;
        var tree = d3.layout.phylotree()
            .svg(d3.select("#tree_display"))
            .options({
                "max-radius":height,
                "zoom":false,
                "show-scale": false,
                // 'left-right-spacing': 'fit-to-size',
                // fit to given size top-to-bottom
                // 'top-bottom-spacing': 'fit-to-size',
                "collapsible": false
            })
            .radial(true)
            .branch_name(function() {
                  return ""
                })



         tree(newick_string);


        var continent2col = {"Europe":"#7fc97f","Asia":"#beaed4","Africa":"#fdc086","South-America":"#ffff99","North-America":"#386cb0","Central-America":"#f0027f","Oceania":"#bf5b17"}
        var tree_attributes = {};
        var maximum_length = 0;

        tree.traverse_and_compute (function (node) {
            if (d3.layout.phylotree.is_leafnode (node)) {
                tree_attributes[node.name] = [tree_meta[node.name]["country"]]
                maximum_length = maximum_length < node.name.length ? node.name.length : maximum_length;
            }
        })


        tree.style_nodes(function(element, node_data){
            if (node_data.name in tree_attributes) {
                var annotation = element.selectAll ("circle").data (tree_attributes[node_data.name]);
                annotation.enter().append ("circle");
                annotation.attr ("r", 3)
                .attr("data-legend",function(d,i) { return d})
                .style ("fill", function(d, i) {
                    return continent2col[d];
                });
            }
        }).layout()


        svg = d3.selectAll("svg")

        var div = d3.select("body").append("div")
            .attr("class", "card tooltip-card")
            .style("opacity", 0);

        var nodes = d3.selectAll(".node")
        nodes.on("mouseover",function(d){
            div.transition()
                .duration(200)
                .style("opacity", .9);
            var htmlString = "<strong>ID:</strong> " +d.name;
            var htmlString = htmlString+"<br><strong>Country:</strong> " + tree_meta[d.name]["country"];
            var htmlString = htmlString+"<br><strong>Date of collection:</strong> " + tree_meta[d.name]["collection_date"];
            div.attr("height", 18*htmlString.split("<br>").length)
            div.html(htmlString)

            .style("left", (d3.event.pageX) + "px")
            .style("top", (d3.event.pageY - 28) + "px");
         })
         .on("mouseout",function(d){
            div.transition()
                .duration(500)
                .style("opacity", 0);
        })
    })
    </script>

{% endblock %}
